src/Common/src/System/Collections/Generic/EnumerableHelpers.cs

```diff
@@ -9,6 +9,63 @@ namespace System.Collections.Generic
     /// <summary>Internal helper functions for working with enumerables.</summary>
     internal static class EnumerableHelpers
     {
+        /// <summary>
+        /// Copies items from an enumerable to an array.
+        /// </summary>
+        /// <typeparam name="T">The element type of the enumerable.</typeparam>
+        /// <param name="source">The source enumerable.</param>
+        /// <param name="array">The destination array.</param>
+        /// <param name="arrayIndex">The index in the array to start copying to.</param>
+        /// <param name="count">The number of items in the enumerable.</param>
+        internal static void Copy<T>(IEnumerable<T> source, T[] array, int arrayIndex, int count)
+        {
+            Debug.Assert(source != null);
+            Debug.Assert(arrayIndex >= 0);
+            Debug.Assert(count >= 0);
+            Debug.Assert(array?.Length - arrayIndex >= count);
+
+            var collection = source as ICollection<T>;
+            if (collection != null)
+            {
+                Debug.Assert(collection.Count == count);
+                collection.CopyTo(array, arrayIndex);
+                return;
+            }
+
+            IterativeCopy(source, array, arrayIndex, count);
+        }
+
+        /// <summary>
+        /// Copies items from a non-collection enumerable to an array.
+        /// </summary>
+        /// <typeparam name="T">The element type of the enumerable.</typeparam>
+        /// <param name="source">The source enumerable.</param>
+        /// <param name="array">The destination array.</param>
+        /// <param name="arrayIndex">The index in the array to start copying to.</param>
+        /// <param name="count">The number of items in the enumerable.</param>
+        internal static void IterativeCopy<T>(IEnumerable<T> source, T[] array, int arrayIndex, int count)
+        {
+            Debug.Assert(source != null && !(source is ICollection<T>));
+            Debug.Assert(arrayIndex >= 0);
+            Debug.Assert(count >= 0);
+            Debug.Assert(array?.Length - arrayIndex >= count);
+
+            using (IEnumerator<T> enumerator = source.GetEnumerator())
+            {
+                for (int i = 0; i < count; i++)
+                {
+                    if (!enumerator.MoveNext())
+                    {
+                        Debug.Fail($"Expected length of {nameof(source)} to be exactly {count}.");
+                    }
+
+                    array[arrayIndex + i] = enumerator.Current;
+                }
+
+                Debug.Assert(!enumerator.MoveNext(), $"Expected length of {nameof(source)} to be exactly {count}.");
```

Comment:
Since this requires that `MoveNext()` will return true exactly `count` times anyway, we could have a conventional `foreach`, leave `count` out of the call and `++arrayIndex` rather than `++i` and doing the addition.
Which of these micro-opts is the greater is hard to guess at from first principles rather than profiling though (and would depend on `MoveNext()` complexity).
Losing the ability to test `count` in debug would make some flaws harder to spot, though they should still be obvious enough to the tests.